test/unit/indentation_test.exs

```diff
@@ -0,0 +1,180 @@
+      "the plunge through" <> # death's whirling vortex, will
+      "a great light" # unfurl over his head?
+      """
+      assert_format_string(bad, good)
+
+      bad = """
+      [{:prev, prev_lineno}] ++
+      [{:prefix_comments, get_prefix_comments(curr_lineno-1, prev_lineno)}] ++
+      [{:prefix_newline, get_prefix_newline(curr_lineno-1, prev_lineno)}] ++ curr_ctx
+      """
+      good = """
+      [{:prev, prev_lineno}] ++
+      [{:prefix_comments, get_prefix_comments(curr_lineno - 1, prev_lineno)}] ++
+      [{:prefix_newline, get_prefix_newline(curr_lineno - 1, prev_lineno)}] ++
+      curr_ctx
+      """
+      assert_format_string(bad, good)
+    end
+  end
+
+  describe "'with' special form indentation" do
+    test "with multiple calls" do
+      bad = """
+      with {:ok, date} <- Calendar.ISO.date(year, month, day),
+           {:ok, time} <- Time.new(hour, minute, second, microsecond),
+           do: new(date, time)
+      """
+      good = """
+      with {:ok, date} <- Calendar.ISO.date(year, month, day),
+           {:ok, time} <- Time.new(hour, minute, second, microsecond) do
+        new(date, time)
+      end
+      """
+      assert_format_string(bad, good)
+
+      assert_format_string """
+      with {year, ""} <- Integer.parse(year),
+           {month, ""} <- Integer.parse(month),
+           {day, ""} <- Integer.parse(day) do
+        new(year, month, day)
+      else
+        _ ->
+          {:error, :invalid_format}
+      end
+      """
+    end
+
+    test "with one call" do
+      assert_format_string("with {:ok, date} <- Calendar.ISO.date(year, month, day), do: new(date, time)\n")
+
+      bad = """
+      with {:ok, date} <- Calendar.ISO.date(year, month, day),
+      do: new(date, time)
+      """
+      good = """
+      with {:ok, date} <- Calendar.ISO.date(year, month, day) do
+        new(date, time)
+      end
+      """
+      assert_format_string(bad, good)
+    end
+
+    test "with messy input" do
+      bad = """
+      with {year, ""} <- Integer.parse(year), {month, ""} <- Integer.parse(month),
+      {day, ""} <- Integer.parse(day) do
+        new(year, month, day)
+      end
+      """
+      good = """
+      with {year, ""} <- Integer.parse(year),
+           {month, ""} <- Integer.parse(month),
+           {day, ""} <- Integer.parse(day) do
+        new(year, month, day)
+      end
+      """
+      assert_format_string(bad, good)
+    end
+  end
+
+  describe "'for' special form indentation" do
+    test "with kw list" do
+      bad = """
+      for app <- apps,
+        do: {app, path},
+        into: %{}
+      """
+      good = """
+      for app <- apps,
+          do: {app, path},
+          into: %{}
+      """
+      assert_format_string(bad, good)
+
+      assert_format_string """
+      for partition <- 0..partitions - 1,
+          pair <- safe_lookup(registry, partition, key),
+          into: [],
+          do: pair
+      """
+    end
```

Comment:
You can write `for` using `do` as well as with a keyword list, worth having a test for that style too?